Cacheo territorial

MariaNavarro90 opened the pull request:
# Información de la Tarea
Vincular el #456

## 📌 Resumen de la Solución
Se implementó un sistema de **cache territorial** para optimizar el manejo de datos provenientes de GESTIONAR, junto con mejoras en validaciones y conteos de prestaciones.

---

## ℹ️ Información Adicional
- Se agregaron nuevos modelos (`TerritorialCache` y `TerritorialSyncLog`).
- Se integró la invalidación del cache con el sistema existente de señales.
- El frontend ahora utiliza un nuevo script para manejar territoriales desde cache.
- Se reforzó la validación de booleanos en serializers.

---

## 🔄 Descripción de los Cambios

### `comedores/models.py`
- **Nuevos modelos:**
  - `TerritorialCache`: guarda territoriales en DB local.
  - `TerritorialSyncLog`: auditoría de sincronizaciones.
- Métodos extra: `esta_desactualizado()`, `to_dict()`.

### `comedores/urls.py`
- Nuevas rutas:
  - `comedores/<int:comedor_id>/territoriales/`
  - `comedores/<int:comedor_id>/territoriales/sincronizar/`
  - `comedores/territoriales/estadisticas/`

### `comedores/services/comedor_service.py`
- Refactor: acceso de `beneficiarios.prestaciones.all()` → `beneficiarios.prestacion`.

### `comedores/templates/comedor/comedor_detail.html`
- Se reemplazó `comedordetail.js` → `comedordetail_territorial_cache.js`.

### `core/cache_utils.py`
- Función `invalidate_territoriales_cache()`.
- Nuevo signal receiver que invalida cache cuando cambia `TerritorialCache`.

### `relevamientos/serializer.py`
- Nuevo método `_convert_yn_to_boolean()`.
- Mejora en `validate()`.
- Logging para debugging de errores de validación.

---

## ✅ Cambios
- [x] Nuevos modelos para cache y logs.
- [x] Nuevas URLs para territoriales.
- [x] Refactor de conteo de prestaciones.
- [x] Nuevo script JS en `comedor_detail.html`.
- [x] Función de invalidación de cache.
- [x] Validación automática de booleanos en serializers.

---

## 🔄 Estrategia de Invalidación

### Invalidación automática por signals
**Archivo:** `core/cache_utils.py:163-166`

```python
@receiver([post_save, post_delete], sender="comedores.TerritorialCache")
def invalidate_territorial_cache_on_change(sender, **kwargs):
    invalidate_territoriales_cache()
```

### Invalidación manual durante sincronización
**Archivo:** `territorial_service.py:212`

- Invalidar cache Django: `cache.delete(cls.CACHE_KEY_TERRITORIALES)`
- Se ejecuta después de cada sync exitoso con GESTIONAR
- Fuerza a recargar datos frescos

### Invalidación por tiempo (TTL)
- **Cache Django:** 1 hora (línea 22)
- **Datos DB considerados obsoletos:** 1 hora (línea 141)
- **Prevención sync frecuentes:** 5 minutos entre syncs (línea 152)

### Invalidación por estado de datos
- Propiedad `esta_desactualizado` en el modelo (`models.py:328-332`)
- Verificación automática de freshness antes de usar cache

### Limpieza administrativa
- Comando `limpiar_cache_completo()` (línea 323)
- Cleanup de logs antiguos (>30 días) via management command

MariaNavarro90 committed:
Agregue los modelos TerritorialCache y TerritorialSyncLog para administrar el almacenamiento en caché de datos territoriales y los registros de sincronización

MariaNavarro90 committed:
Agregue endpoints para la gestión de datos territoriales en la API

MariaNavarro90 committed:
Agregue vistas para la gestión de territoriales con caché, incluyendo endpoints para obtener territoriales, forzar sincronización y estadísticas del caché.

MariaNavarro90 committed:
Agregue comando de gestión para sincronizar el caché de territoriales, incluyendo opciones para forzar sincronización, limpiar logs antiguos y mostrar estadísticas del caché.

MariaNavarro90 committed:
Migracion

MariaNavarro90 committed:
Optimizar el cálculo de conteo de prestaciones al obtener directamente la prestación del beneficiario en lugar de iterar sobre todas las prestaciones.

MariaNavarro90 committed:
Agregado servicio para gestionar el caché de territoriales con sincronización desde GESTIONAR

MariaNavarro90 committed:
Eliminación de comentarios innecesarios en el servicio de gestión de caché de territoriales para mejorar la legibilidad del código.

MariaNavarro90 committed:
Actualizar script de detalle de comedor para utilizar caché territorial

MariaNavarro90 committed:
Agregada la función para invalidar cache de territoriales y su señal de invalidación automática

MariaNavarro90 committed:
Agregar validación personalizada y conversión de valores Y/N a booleanos en RelevamientoSerializer

MariaNavarro90 committed:
Agregar sistema de caché híbrido para territoriales con indicadores de estado y sincronización manual

MariaNavarro90 committed:
Merge branch 'development' of https://github.com/dsocial118/SISOC-Backoffice into CacheoTerritorial

MariaNavarro90 committed:
Merge branch 'development' of https://github.com/dsocial118/SISOC-Backoffice into CacheoTerritorial

MariaNavarro90 committed:
black

MariaNavarro90 committed:
pylint

MariaNavarro90 committed:
Refactor del menu lateral con indentacion

MariaNavarro90 committed:
Merge branch 'development' of https://github.com/dsocial118/SISOC-Backoffice into CacheoTerritorial

MariaNavarro90 committed:
Mejorado el modelo TerritorialCache: agregado campo de provincia, actualizado el orden y mejorada la representación de cadenas

MariaNavarro90 committed:
Migraciones

MariaNavarro90 committed:
Se corrigió referencia de variable en ComedorService para usar relevamiento en lugar de beneficiarios para prestación

MariaNavarro90 committed:
Mejora en el servicio TerritorialService: implementación de gestión de cache por provincia, optimización de la estrategia de sincronización y mejoras en la UX al mostrar datos en caché mientras se realiza la sincronización.

MariaNavarro90 committed:
Añadida función para invalidar cache territorial por provincia y mejorada la gestión de invalidación en cambios de datos territoriales.

MariaNavarro90 committed:
Mejora en la función convert_to_boolean: se agregó manejo para valores booleanos y strings adicionales, mejorando la robustez y claridad del código.

MariaNavarro90 committed:
Refactorización del manejo de errores en RelevamientoApiView: se movió la llamada al método clean() a un bloque try-except para mejorar la gestión de errores durante el procesamiento de datos.

MariaNavarro90 committed:
black .

MariaNavarro90 committed:
djlint

MariaNavarro90 committed:
Pylint

juanikitro commented on comedores/services/territorial_service.py:
```diff
@@ -0,0 +1,576 @@
+            # Verificar último sync exitoso para esta provincia
+            ultimo_sync = (
+                TerritorialSyncLog.objects.filter(
+                    exitoso=True, comedor__provincia_id=provincia_id
```
Bloqueante: Esto va a fallar ya que `comedor_id` es un int y no una fk

juanikitro commented on comedores/services/territorial_service.py:
```diff
@@ -0,0 +1,576 @@
+        """Limpia todo el cache de territoriales."""
+        try:
+            # Limpiar Django cache
+            cache.delete(cls.CACHE_KEY_TERRITORIALES)
```
Bloqueante: el cache de la provincia nunca se va a limpiar porque `CACHE_KEY_TERRITORIALES` no veo que exista. Puede ser que me equivoque, chequear por favor

MariaNavarro90 replied:
Confirmo que tenes razón: En la línea
  509, el método `limpiar_cache_completo()`
  intenta limpiar el cache usando:

  `cache.delete(cls.CACHE_KEY_TERRITORIALES)`

  Pero revisando todo el archivo,
  `CACHE_KEY_TERRITORIALES` no está definido 
  como constante de clase.

  Lo que sí existe:

  - Línea 27-29:
  `_get_cache_key_provincia(provincia_id)`
  que genera claves dinámicas por provincia
   ```(f"territoriales_provincia_{provincia_id
  }")```
  - Constantes definidas: Solo
  `CACHE_TIMEOUT y SYNC_TIMEOUT`

juanikitro commented on comedores/services/territorial_service.py:
```diff
@@ -0,0 +1,576 @@
+        return territoriales
+
+    @classmethod
+    def _actualizar_cache_db(cls, territoriales_data: List[Dict]) -> int:
```
Duda: este metodo no actualiza nada en la DB. Mas alla de eso, gestionar_uid y nombre no van a ninguna lado. No entiendo de que sirve mas que para contar

juanikitro commented on comedores/services/territorial_service.py:
```diff
@@ -0,0 +1,576 @@
+        return contador
+
+    @classmethod
+    def _actualizar_cache_db_provincia(
```
Sugerencia: Añadiria un `transaction.atomic()` porque se hacen muchos creates pero si algo falla que dara a medias

juanikitro commented on comedores/services/territorial_service.py:
```diff
@@ -0,0 +1,576 @@
+                territoriales_ejemplo = cls._obtener_desde_db_por_provincia(
+                    provincia_id
+                )
+                if territoriales_ejemplo["territoriales"]:
```
Sugerencia: Borrar todo lo que traiga territoriales de ejemplo. Si no se conecta se debe usar el cache. Si no tenemos el cache que mande vacio, si no estariamos mintiendo

juanikitro commented on comedores/management/commands/sync_territoriales_cache.py:
```diff
@@ -0,0 +1,182 @@
+                    comedor_id=comedor.id, forzar_sync=force
+                )
+
+                if resultado["fuente"] in [
```
Duda: Esto funcionaria? Tenemos esas fuentes?

juanikitro commented on comedores/management/commands/sync_territoriales_cache.py:
```diff
@@ -0,0 +1,182 @@
+        # Obtener muestra de comedores activos
+        comedores_muestra = Comedor.objects.filter(
+            estado__in=["Activo", "Asignado a Dupla Técnica", "En proceso"]
+        ).order_by("?")[
```
Sugerencia: Esto esta bien implementado, pero si tenemos muchos comedores va a ser mucha carga al pedo para el SQL. Se podria optimizar un poco. No es obligatorio, pero solo para pensar

MariaNavarro90 committed:
Merge branch 'development' of https://github.com/dsocial118/SISOC-Backoffice into CacheoTerritorial

MariaNavarro90 committed:
Comentario de PR sobre CACHE_KEY_TERRITORIALES

Eliminado: cache.delete(cls.CACHE_KEY_TERRITORIALES) (clave inexistente)
Agregado: Iteración sobre todas las provincias para limpiar sus claves de cache individuales

MariaNavarro90 committed:
Comentario PR: Bloqueante: Esto va a fallar ya que comedor_id es un int y no una fk

Primera query: Obtiene lista de IDs de comedores de la provincia
Segunda query: Filtra logs donde comedor_id esté en esa lista
Resultado: Funciona correctamente sin intentar hacer join inexistente

MariaNavarro90 committed:
Comentario PR: "Sugerencia: Añadiria un transaction.atomic() porque se hacen muchos creates pero si algo falla quedara a medias" Se agrego transaction.atomic() al método

MariaNavarro90 committed:
Comentario de PR:
Sugerencia: Borrar todo lo que traiga territoriales de ejemplo. Si no se conecta se debe usar el cache. Si no tenemos el cache que mande vacio, si no estariamos mintiendo

Refactorizada la lógica de recuperación de datos territoriales para eliminar la creación de datos de ejemplo heredados y mejorado el registro de errores para el servicio GESTIONAR no disponible.

MariaNavarro90 committed:
Comentario de PR: Duda: Esto funcionaria? Tenemos esas fuentes?

Se corrigieron las funtes

MariaNavarro90 committed:
Comentario de PR: Sugerencia: Esto esta bien implementado, pero si tenemos muchos comedores va a ser mucha carga al pedo para el SQL. Se podria optimizar un poco. No es obligatorio, pero solo para pensar

Optimizado el muestreo aleatorio de comedores activos para el proceso de sincronización.
Sin ORDER BY aleatorio: No ordena toda la tabla
Usa LIMIT/OFFSET: Solo trae los registros necesarios

MariaNavarro90 committed:
black

dsocial118 committed:
Merge branch 'development' into CacheoTerritorial
